src/librustc_mir/transform/util.rs

```diff
@@ -49,3 +50,70 @@ pub fn retain_basic_blocks(mir: &mut Mir, keep: &[bool]) {
 
     update_basic_block_ids(mir, &replacements);
 }
+
+// A simple map to perform quick lookups of the predecessors of a BasicBlock.
+// Since BasicBlocks usually only have a small number of predecessors, we use a
+// simple vector. Also, if a block has the same target more than once, for
+// example in a switch, it will appear in the target's predecessor list multiple
+// times. This allows to update the map more easily when modifying the graph.
+pub struct PredecessorMap {
+    map: Vec<Vec<BasicBlock>>,
```

Comment:
As far as I can tell, this pass seems to only care about how many predecessors a block has, not what those predecessors are. It would be simpler and more efficient to just store a `Vec<usize>`.